Refactor to use Mdbook and introduction (#4)
mun-lang/book: src/ch01-00-types.md

# Types

> Mun is a statically typed languages which helps to detect type-related errors in

Reviewer: ```suggestion
Mun is a statically typed language, which helps to detect type-related errors at
```

> a program at compile-time. A type error is an invalid operation on a given type

Reviewer: ```suggestion
compile-time. A type error is an invalid operation on a given type, such as an
```

> such as a division by a string, trying to access a field that doesn't exist or

Reviewer: ```suggestion
integer divided by a string, trying to access a field that doesn't exist, or
```

calling a function with the wrong number of arguments.

Some languages require a programmer to explicitly annotate syntactic constructs
with type information:

```c++

> int foo = 3+4;

Reviewer: ```suggestion
int foo = 3 + 4;
```

```

> Often the types of variables can be inferred by their usage just as in the

Reviewer: ```suggestion
However, often variable types can be inferred by their usage, as illustrated
```

> example above. Mun uses type inferencing to determine the type of a variable at

Reviewer: ```suggestion
above. Mun uses type inferencing to determine variable types at compile time.
```

> compile time. However, it does force you to explicitly annotate variables in a

Reviewer: ```suggestion
You are still forced to explicitly annotate variables in a few locations to ensure
```

> few locations to ensure a contract between interdependent code.

Reviewer: ```suggestion
a contract between interdependent code.
```

```mun
function bar(a:int) {

>     let foo = 3+a

Reviewer: ```suggestion
    let foo = 3 + a
```

}
```

Here, the parameter `a` must be annotated because it solidifies the signature of

> the function. However, the type of `foo` can be inferred through it's usage.

Reviewer: ```suggestion
the function. The type of `foo` can be inferred through its usage.
```

## Numeric types

Mun knows two basic numeric types: `float` and `int`. A `float` is a
double-precision IEEE 64-bit floating point number and an `int` represents a
64-bit integral number.

> In Mun an `int` can be explicitly casted to `float` but the reverse is not true:

Reviewer: ```suggestion
In Mun an `int` can be explicitly cast to `float`, but the reverse is not true.
```

Reviewer: Do you mean `implicitly cast`? Otherwise the next line should be changed to read `and can therefore not be cast explicitly`.

> assigning a `float` to an `int` might cause loss of precision and is therefor

Reviewer: ```suggestion
Assigning a `float` to an `int` might cause loss of precision and can therefore
```

> not allowed implicitly.

Reviewer: ```suggestion
not be cast implicitly.
```

Reviewer: `implicit` or `explicit`?
